File: DD_c_plusplus/Examples/KnapsackInstance/KnapsackGurobiClass.cpp
#include "KnapsackGurobiClass.h"

KnapsackGurobiClass::KnapsackGurobiClass(const KnapsackInstance& params, bool continuous)
        : AbstractProblemGurobi(params.variables.size(), params.objective_weights, continuous),
          weights(params.weights),
          capacity(params.right_side_of_restrictions)
        {}


void KnapsackGurobiClass::add_constraints() {
    // (1) Capacity constraint: total weight of selected items must not exceed capacity
    GRBLinExpr expr = 0;
    for (int a = 0; a < variable_length; ++a) {
        expr += weights[a] * x[a];
    }
    model->addConstr(expr <= capacity, "Capacity");
}

void KnapsackGurobiClass::show_results() {
    if (model->get(GRB_IntAttr_Status) == GRB_OPTIMAL) {
        cout << "\n------------------------------" << endl;
        cout << "Optimal total value: " << model->get(GRB_DoubleAttr_ObjVal) << endl;
        cout << "Items included in the knapsack: " << endl;
        for (int a = 0; a < variable_length; ++a) {
            if (x[a].get(GRB_DoubleAttr_X) != 0) {
                cout << "Article " << a << " - Weight: " << weights[a] << ", Benefit: " << objective_weights[a] << ", Fraction: " << x[a].get(GRB_DoubleAttr_X) << endl;
            }
        }
    } else {
        cout << "No optimal solution found." << endl;
    }
}

File: DD_python/Examples/KnapsackInstance/KnapsackGurobiClass.py
from gurobipy import GRB, quicksum
from Examples.AbstractGurobiProblems.AbstractProblemGurobiClass import AbstractProblemGurobi
from Examples.KnapsackInstance.KnapsackInstance import KnapsackStructure

class KnapsackGurobi(AbstractProblemGurobi):
    '''
    SourceCode to solve the knapsack problem using Gurobi.
    '''
    def __init__(self, parameters: 'KnapsackStructure', continuous: bool=False) -> None:
        '''
        Initialize the KnapsackGurobi instance.

        Parameters:
        - parameters (KnapsackStructure): Instance of the knapsack problem, containing:
            * variables (list[tuple[str, list[int]]]): variable names and their domains.
            * weights (list[int]): weights of each item for the capacity constraint.
            * objective_weights (list[int]): profits/values of each item.
            * right_side_of_restrictions (int): maximum capacity of the knapsack.
        - continuous (bool): Indicates if the variables should be continuous (True) or binary (False).
        '''
        super().__init__(variable_length=len(parameters.variables),
                         objective_weights=parameters.objective_weights,
                         continuous=continuous,
                         min_max=GRB.MAXIMIZE)

        self.params: 'KnapsackStructure' = parameters
        self.weights: list[float] = parameters.weights
        self.capacity: float = parameters.right_side_of_restrictions

    def _add_constraints(self) -> None:
        # (1) Capacity constraint: total weight of selected items must not exceed capacity
        self.model.addConstr(quicksum(self.weights[i]*self.x[i] for i in self.set_variable_length) <= self.capacity, "Capacity")

    def show_results(self) -> None:
        '''
        Show the results of the model optimization.
        '''
        if self.model.status == GRB.OPTIMAL:
            print("\n------------------------------")
            print(f"Optimal total value:  {self.model.objVal:.6g}")
            print("Items included in the knapsack: ")
            for i in self.set_variable_length:
                if self.x[i].x != 0:
                    print(f'Article {i} - Weight:  {self.weights[i]}, Benefit: {self.objective_weights[i]}, Fraction:  {self.x[i].x:.6g}')
        else:
            print("No optimal solution found.")
